### PR TITLE
chore(broker-core): close subscription when event scope is left

### LINKED ISSUE #1016: Close a subscription when the event scope is left
AT:
* close a subscription when the event scope is left (e.g. scope of the intermediate message catch event is compleded/canceled)
* send the close request over the message endpoint as "remote command" (single message, ACKs, retry)

Context:
* when the subscription is closed then no message can be correlated to this subscription anymore

### LINKED ISSUE #1166: Close a subscription when a message is correlated
AT:
* a message subscription is closed when the first message is correlated

Context:
* currently, we don't support non-interrupting message events
* so, when the first message is correlated then the subscription can be closed 
* this is related to #1016

### PR DESCRIPTION
**Pull request by saig0**
* new subscription command CLOSE
* send CLOSE command when the message catch element is terminated
* send the command again if not acknowledged

closes #1016

### CODE REVIEW

**Comment by saig0** [comment]
Name `activityInstanceKey` as `elementInstanceKey` as part of #1536

File: broker-core/src/test/java/io/zeebe/broker/workflow/MessageCorrelationTest.java [hunk 1]
```diff
@@ -480,39 +484,53 @@ public void shouldCorrelateMessageSubscription() {
...
-        .send();
+    testClient.cancelWorkflowInstance(workflowInstanceKey);
 
-    testClient.receiveElementInState("wf", WorkflowInstanceIntent.ELEMENT_TERMINATED);
+    final SubscribedRecord messageSubscription =
+        findMessageSubscription(MessageSubscriptionIntent.CLOSED);
+    assertThat(messageSubscription.valueType()).isEqualTo(ValueType.MESSAGE_SUBSCRIPTION);
+    assertThat(messageSubscription.recordType()).isEqualTo(RecordType.EVENT);
+    assertThat(messageSubscription.value())
+        .containsExactly(
```
Comment (Zelldon) [inline broker-core/src/test/java/io/zeebe/broker/workflow/MessageCorrelationTest.java]: Please change to Recordingexporter

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]

File: broker-core/src/main/java/io/zeebe/broker/subscription/command/SubscriptionApiCommandMessageHandler.java [hunk 2]
```diff
@@ -180,6 +194,38 @@ private boolean onCorrelateMessageSubscription(DirectBuffer buffer, int offset,
...
+    return writeCommand(
+        closeMessageSubscriptionCommand.getSubscriptionPartitionId(),
+        ValueType.MESSAGE_SUBSCRIPTION,
+        MessageSubscriptionIntent.CLOSE,
+        messageSubscriptionRecord);
+  }
+
+  private boolean onCloseWorkflowInstanceSubscription(DirectBuffer buffer, int offset, int length) {
+    closeWorkflowInstanceSubscriptionCommand.wrap(buffer, offset, length);
+
```
[the remark this answers is not in the document]
Comment (saig0) [reply broker-core/src/main/java/io/zeebe/broker/subscription/command/SubscriptionApiCommandMessageHandler.java]: Good point.

File: broker-core/src/test/java/io/zeebe/broker/workflow/MessageCorrelationTest.java (hunk 1 above)
Comment (saig0) [reply broker-core/src/test/java/io/zeebe/broker/workflow/MessageCorrelationTest.java]: Agree.

**Comment by saig0** [comment]
@Zelldon  I applied the review hints.

**Review by ChrisKujawa — APPROVED** [review]
bors r+

**Comment by menski** [comment]
bors retry

**Pull request merged**